Tasks/CondaEnvironment/conda_internal.ts

```diff
@@ -0,0 +1,141 @@
+import * as fs from 'fs';
+import * as path from 'path';
+
+import * as task from 'vsts-task-lib/task';
+import * as tool from 'vsts-task-tool-lib/tool';
+import { ToolRunner } from 'vsts-task-lib/toolrunner';
+
+import { Platform } from './taskutil';
+
+/**
+ * Whether `searchDir` has a `conda` executable for `platform`.
+ * @param searchDir Absolute path to a directory to look for a `conda` executable in.
+ * @param platform Platform whose executable type we want to use (i.e. `conda` vs `conda.exe`)
+ * @returns Whether `searchDir` has a `conda` executable for `platform`.
+ */
+export function hasConda(searchDir: string, platform: Platform): boolean {
+    const conda = platform === Platform.Windows ?
+        path.join(searchDir, 'Scripts', 'conda.exe') :
+        path.join(searchDir, 'bin', 'conda');
+
+    return fs.existsSync(conda) && fs.statSync(conda).isFile();
+}
+
+/**
+ * Add Conda's `python` and `conda` executables to PATH.
+ * Precondition: Conda is installed at `condaRoot`
+ * @param condaRoot Root directory or "prefix" of the Conda installation
+ * @param platform Platform for which Conda is installed
+ */
+export function prependCondaToPath(condaRoot: string, platform: Platform): void {
+    if (platform === Platform.Windows) {
+        // Windows: `python` lives in `condaRoot` and `conda` lives in `condaRoot\Scripts`
+        tool.prependPath(condaRoot);
+        tool.prependPath(path.join(condaRoot, 'Scripts'));
+    } else {
+        // Linux and macOS: `python` and `conda` both live in the `bin` directory
+        tool.prependPath(path.join(condaRoot, 'bin'));
+    }
+}
+
+/**
+ * Download the appropriate Miniconda installer for `platform`.
+ * @param platform Platform for which we want to download Miniconda.
+ * @returns Absolute path to the download.
+ */
+export function downloadMiniconda(platform: Platform): Promise<string> {
+    const url = (() => {
+        switch (platform) {
+            case Platform.Linux: return 'https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh';
+            case Platform.MacOS: return 'https://repo.continuum.io/miniconda/Miniconda3-latest-MacOSX-x86_64.sh';
+            case Platform.Windows: return 'https://repo.continuum.io/miniconda/Miniconda3-latest-Windows-x86_64.exe';
+        }
+    })();
+
+    // By default `downloadTool` will name the downloaded file with a GUID
+    // But on Windows, the file must end with `.exe` to make it executable
+    const tempDirectory = task.getVariable('AGENT_TEMPDIRECTORY');
```

Comment:
I think it's actually `AGENT.TEMPDIRECTORY` (at least that seems to be what most other tasks use... I think any '.' get converted to '_' under the covers.  Only the tests seem to use '_'.)

Reply:
Both will work

Reply:
I'm not sure both are 'correct' the only other tasks that use `_` are those that by-pass the framework and directly read the environment variable.

Reply:
The dot notation seems more VSTS native to me. For the dot notation, `Agent.TempDirectory` is the canonical casing. Interesting that this variable isn't documented at https://docs.microsoft.com/en-us/vsts/build-release/concepts/definitions/build/variables

Reply:
Fair enough.  I'll change it to use the dot.

For what it's worth, the `Agent.ToolsDirectory` variable isn't documented either.